torchtune/modules/transformer.py

```diff
@@ -90,13 +99,147 @@ def forward(
+        """Check if the key value caches are setup."""
+        return self.attn.kv_cache is not None
+
+    def reset_cache(self):
+        """Reset the key value caches."""
+        self.attn.reset_cache()
+
+    def _skip_mask(self, mask: Optional[Tensor]) -> Optional[Tensor]:
+        """Some tokens in x may not attend to any encoder inputs
+        due to the cross attention mask (encoder_mask). This results in
+        a full row of the attention matrix being masked out.
+
+        In the example below, the word "the" is masked from every embedding.
+
+        .. code-block:: text
+
+            |emb||emb||emb|
+        |The| x    x    x
+        |red|      x
+        |car| x
+
+        This results in no inputs into the softmax layer which causes a NaN.
+        The skip mask removes the output of the attention module and
+        mlp resulting in the token being skipped.
+
+        """
+        if mask is None:
+            return None
+        if mask.dtype == torch.bool:
+            mask = ~mask
+        else:
+            mask = torch.isneginf(mask)
+        mask = torch.all(mask, dim=-1, keepdim=True)
+        return mask
+
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        **kwargs: Dict,
+    ) -> Tensor:
+        """
+        Args:
+            x (Tensor): input tensor with shape
+                [batch_size x seq_length x embed_dim]
+            encoder_input (Optional[Tensor]): Optional second input to cross attend with x. Shape
+                [batch_size x seq_length x embed_dim] (seq_length and embed_dim may very from x)
+            encoder_mask (Optional[Tensor]):  Cross attention boolean tensor with shape
+                [batch_size x x_seq_len x encoder_seq_len]
+            **kwargs (Dict): transformer layer inputs not relevant to self attention.
+
+        Returns:
+            Tensor: output tensor with same shape as input
+                [batch_size x seq_length x embed_dim]
+        """
+        # During decoding, it's possible encoder_input is None because the embeds
+        # are already stored in the kv cache.
+        empty_cache = not self.cache_enabled or self.attn.kv_cache.size == 0
+        # Skip cross attention when no secondary input as it's primary purpose
+        # is to attend between x and encoder_input.
+        if encoder_input is None and empty_cache:
+            return x
+
+        # A mask of tokens (x) with no encoder_input
+        skip_mask = self._skip_mask(encoder_mask)
+
+        # Input tensor and attention output have the same shape
+        # [b, s, d]
+        # Norm applied before self-attention
+        # TODO: Add support for sample packing and bring ack input_pos
+        attn_out = self.attn(self.ca_norm(x), encoder_input, mask=encoder_mask)
+        if skip_mask is not None:
+            attn_out.masked_fill_(skip_mask, 0)
```

Comment:
I don't think I follow?

Reply:
skip_mask represents rows in the original mask that are fully masked out, i.e a given token can't attend to any encoder embeds. This results in NaN values in that row from attention. This replaces them with 0s in place.